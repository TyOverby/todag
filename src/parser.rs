use itertools::{put_back, PutBack};
use pulldown_cmark::{Event, Options, Parser, Tag};
use Event::*;

#[derive(Debug)]
pub struct Task<'a> {
    pub checked: bool,
    pub path: Vec<String>,
    pub name: Vec<Event<'a>>,
    pub description: Vec<Event<'a>>,
    pub nested: Vec<Task<'a>>,
}

#[derive(Debug)]
pub struct Header {
    depth: i32,
    text: String,
}

#[derive(Copy, Clone)]
enum TaskParseState {
    WaitingForTaskListMarker,
    ParsingTitle,
    ParsingDescription,
}

use TaskParseState::*;

fn parse_tasks<'a>(out: &mut Vec<Task<'a>>, paths: &[Header], parser: &mut PutBack<Parser<'a>>) {
    while let Some(event) = parser.next() {
        match event {
            Start(Tag::Item) => out.push(parse_task(paths, parser)),
            End(Tag::List(_)) => break,
            other => eprintln!("{:?} not implemented", other),
        }
    }
}

fn parse_task<'a>(paths: &[Header], parser: &mut PutBack<Parser<'a>>) -> Task<'a> {
    let mut checked = false;
    let mut name = Vec::new();
    let mut description = Vec::new();
    let path = paths.iter().map(|header| &header.text).cloned().collect();
    let mut state = WaitingForTaskListMarker;
    let mut nested = vec![];
    while let Some(event) = parser.next() {
        match (event, state) {
            (TaskListMarker(c), WaitingForTaskListMarker) => {
                checked = c;
                state = ParsingTitle;
            }
            (SoftBreak, ParsingTitle) => state = ParsingDescription,
            (event @ End(Tag::Item), ParsingTitle) => {
                parser.put_back(event);
                state = ParsingDescription;
            }
            (Start(Tag::List(_)), ParsingTitle) | 
            (Start(Tag::List(_)), ParsingDescription)=> {
                parse_tasks(&mut nested, paths, parser);
                state = ParsingDescription;
            }
            (event, ParsingTitle) => name.push(event),
            (event @ End(Tag::Item), ParsingDescription) => {
                parser.put_back(event);
                break;
            }
            (event, ParsingDescription) => description.push(event),
            _ => break,
        }
    }
    Task {
        path,
        name,
        checked,
        description,
        nested,
    }
}

fn parse_header(depth: i32, parser: &mut PutBack<Parser>) -> Header {
    let mut text = String::new();
    while let Some(event) = parser.next() {
        match event {
            Event::Text(t) => text.push_str(&t),
            Event::End(Tag::Header(_)) => break,
            _ => break,
        }
    }
    Header { depth, text }
}

pub fn extract(s: &str) -> (Vec<Task>, Vec<Header>) {
    let mut options = Options::empty();
    options.insert(Options::ENABLE_TASKLISTS);

    let mut parser = put_back(Parser::new_ext(s, options));
    let mut out = Vec::new();
    let mut path_stack: Vec<Header> = Vec::new();
    while let Some(event) = parser.next() {
        match event {
            Event::Start(Tag::Header(i)) => {
                let header = parse_header(i, &mut parser);
                while let Some(top_header) = path_stack.pop() {
                    if top_header.depth < header.depth {
                        path_stack.push(top_header);
                        break;
                    }
                }
                path_stack.push(header);
            }
            Event::Start(Tag::List(None)) => {
                parse_tasks(&mut out, &path_stack, &mut parser);
            }
            other => {
                eprintln!("{:?} not implemented", other);
            }
        }
    }

    (out, path_stack)
}
