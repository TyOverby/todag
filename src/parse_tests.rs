use crate::parser::extract;

mod debug_output {
    use pulldown_cmark::{Options, Parser};

    #[test]
    fn single_tasks_debugout() {
        use pulldown_cmark::{Options, Parser};
        let mut options = Options::empty();
        options.insert(Options::ENABLE_TASKLISTS);
        let input = r#"
- [ ] this is the title 
      this is the description
"#;
        let v: Vec<_> = Parser::new_ext(input, options).collect();

        insta::assert_debug_snapshot!(v);
    }

    #[test]
    fn nested_tasks_with_weird_indentation_test() {
        use pulldown_cmark::{Options, Parser};
        let mut options = Options::empty();
        options.insert(Options::ENABLE_TASKLISTS);
    let input = r#"
- [ ] title of parent
  description of parent
  - [ ] title of child
    description of child
  more description of parent
"#;
        let v: Vec<_> = Parser::new_ext(input, options).collect();

        insta::assert_debug_snapshot!(v);
    }


    #[test]
    fn nested_tasks_debugout() {
        let mut options = Options::empty();
        options.insert(Options::ENABLE_TASKLISTS);
        let input = r#"
- [ ] a 
  - [ ] b
    - [ ] c 
  - [ ] d
"#;
        let v: Vec<_> = Parser::new_ext(input, options).collect();

        insta::assert_debug_snapshot!(v);
    }

    #[test]
    fn two_tasks_debugout() {
        let mut options = Options::empty();
        options.insert(Options::ENABLE_TASKLISTS);
        let input = r#"
- [ ] a name
      a description
- [ ] b name
      b description
"#;
        let v: Vec<_> = Parser::new_ext(input, options).collect();
        insta::assert_debug_snapshot!(v);
    }
}

#[test]
fn nested_task() {
    let input = r#"
- [ ] title of parent
  - [ ] title of child
"#;
    let (tasks, _headers) = extract(input);
    insta::assert_debug_snapshot!(tasks);
}

#[test]
fn nested_tasks_with_descriptions() {
    let input = r#"
- [ ] title of parent
      description of parent
  - [ ] title of child
        description of child
"#;
    let (tasks, _headers) = extract(input);
    insta::assert_debug_snapshot!(tasks);
}

#[test]
fn deeply_nested_task() {
    let input = r#"
- [ ] a 
  - [ ] b
    - [ ] c 
  - [ ] d
"#;
    let (tasks, _headers) = extract(input);
    insta::assert_debug_snapshot!(tasks);
}

#[test]
fn single_checkbox_with_title_unchecked() {
    let input = r#"
- [ ] this is the title"#;
    let (tasks, _headers) = extract(input);
    insta::assert_debug_snapshot!(tasks);
}

#[test]
fn two_checkbox_with_description() {
    let input = r#"
- [ ] a name
      a description
- [ ] b name
      b description
"#;
    let (tasks, _headers) = extract(input);
    insta::assert_debug_snapshot!(tasks);
}

#[test]
fn two_checkbox_without_description() {
    let input = r#"
- [ ] a name
- [ ] b name
"#;
    let (tasks, _headers) = extract(input);
    insta::assert_debug_snapshot!(tasks);
}

#[test]
fn single_checkbox_with_description() {
    let input = r#"
- [ ] this is the title
      and this is the description
"#;
    let (tasks, _headers) = extract(input);
    insta::assert_debug_snapshot!(tasks);
}

#[test]
fn single_checkbox_with_title_checked() {
    let input = r#"
- [x] this is the title"#;
    let (tasks, _headers) = extract(input);
    insta::assert_debug_snapshot!(tasks);
}

#[test]
fn single_header() {
    let input = r#"
# a "#;
    let (_tasks, headers) = extract(input);
    insta::assert_debug_snapshot!(headers);
}

#[test]
fn task_under_header() {
    let input = r#"
# Header Name
- [ ] Task Name
"#;
    let (tasks, _headers) = extract(input);
    insta::assert_debug_snapshot!(tasks);
}

#[test]
fn chained_header() {
    let input = r#"
# a
## b
### c "#;
    let (_tasks, headers) = extract(input);
    insta::assert_debug_snapshot!(headers);
}
