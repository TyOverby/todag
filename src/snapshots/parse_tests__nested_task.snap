---
source: src/parse_tests.rs
expression: tasks
---
[
    Task {
        checked: false,
        path: [],
        name: [
            Text(
                Borrowed(
                    "title of parent",
                ),
            ),
        ],
        description: [],
        nested: [
            Task {
                checked: false,
                path: [],
                name: [
                    Text(
                        Borrowed(
                            "title of child",
                        ),
                    ),
                ],
                description: [],
                nested: [],
            },
        ],
    },
]
